File: server/player.js
/** Authoritative player: stats, pack, bank, equipment, skills. */
import { SPAWN, INV_SIZE, BANK_SIZE, EQUIP_SLOTS, uid, ACTION, xpToLevel, hueFromName } from '../shared/constants.js';
import { pack, S2C } from '../shared/protocol.js';
import { item, playerStats, skillLevel } from '../shared/catalog.js';
import {
  emptyInv, addItem as invAdd, removeItem as invRemove, countOf, canFit,
  serializeInv,
} from './systems/inventory.js';

export class Player {
  constructor(ws, name) {
    this.id = uid('p');
    this.ws = ws;
    this.name = name;
    this.x = SPAWN.x;
    this.y = SPAWN.y;
    this.dir = 0;
    this.path = [];
    this.hp = 40;
    this.maxHp = 40;
    this.inv = emptyInv(INV_SIZE);
    this.bank = emptyInv(BANK_SIZE);
    this.equip = { weapon: null, armor: null, charm: null };
    this.skills = { combat: { xp: 0 }, foraging: { xp: 0 }, binding: { xp: 0 } };
    this.action = ACTION.IDLE;
    this.target = null;
    this.tradeId = null;
    this.intent = null;
    this.channel = null;
    this.buff = null;
    this.lastSwing = 0;
    this.atk = 4;
    this.def = 2;
    this.joinedAt = Date.now();
    this.lastChat = 0;
    this.regenAcc = 0;
    this.hue = hueFromName(name);
  }

  send(type, payload) {
    if (this.ws && this.ws.readyState === 1) {
      try { this.ws.send(pack(type, payload)); } catch { /* closed */ }
    }
  }

  addItem(id, qty) { return invAdd(this.inv, id, qty); }
  removeItem(id, qty) { return invRemove(this.inv, id, qty); }
  countItem(id) { return countOf(this.inv, id); }
  hasItems(list) {
    for (const { id, qty } of list) if (countOf(this.inv, id) < qty) return false;
    return true;
  }
  canFit(id, qty) { return canFit(this.inv, id, qty); }

  bonusAtk() {
    let n = 0;
    for (const slot of EQUIP_SLOTS) {
      const it = this.equip[slot] ? item(this.equip[slot]) : null;
      if (it) n += it.atk || 0;
    }
    return n;
  }
  bonusDef() {
    let n = 0;
    for (const slot of EQUIP_SLOTS) {
      const it = this.equip[slot] ? item(this.equip[slot]) : null;
      if (it) n += it.def || 0;
    }
    return n;
  }
  bonusHp() {
    let n = 0;
    for (const slot of EQUIP_SLOTS) {
      const it = this.equip[slot] ? item(this.equip[slot]) : null;
      if (it) n += it.hp || 0;
    }
    return n;
  }

  refresh() {
    const s = playerStats(this);
    const ratio = this.maxHp > 0 ? this.hp / this.maxHp : 1;
    this.maxHp = s.maxHp;
    this.atk = s.atk;
    this.def = s.def;
    this.hp = Math.max(0, Math.min(s.maxHp, Math.round(ratio * s.maxHp)));
  }

  addXp(skillId, amount) {
    if (!this.skills[skillId]) this.skills[skillId] = { xp: 0 };
    const before = skillLevel(this, skillId);
    this.skills[skillId].xp += Math.max(0, amount | 0);
    const after = skillLevel(this, skillId);
    const prog = xpToLevel(this.skills[skillId].xp);
    this.send(S2C.XP, {
      skill: skillId, xp: this.skills[skillId].xp, amount, level: after,
      into: prog.into, need: prog.need,
    });
    if (after > before) {
      this.refresh();
      this.hp = Math.min(this.maxHp, this.hp + 8);
      this.send(S2C.NOTIFY, { text: `${labelSkill(skillId)} rises to ${after}.`, kind: 'xp' });
    }
    return after;
  }

  publicView() {
    return {
      id: this.id,
      name: this.name,
      x: r2(this.x),
      y: r2(this.y),
      hp: Math.round(this.hp),
      maxHp: this.maxHp,
      action: this.action,
      dir: r2(this.dir),
      hue: this.hue,
    };
  }

  privateView() {
    return {
      ...this.publicView(),
      inv: serializeInv(this.inv),
      bank: serializeInv(this.bank),
      equip: { weapon: this.equip.weapon, armor: this.equip.armor, charm: this.equip.charm },
      skills: {
        combat: this.skills.combat.xp,
        foraging: this.skills.foraging.xp,
        binding: this.skills.binding.xp,
      },
      atk: this.atk,
      def: this.def,
      target: this.target,
      coins: this.countItem('ember_coin'),
      channel: this.channel ? { type: this.channel.type, until: this.channel.until } : null,
    };
  }
}

function labelSkill(id) {
  return { combat: 'Ashen Combat', foraging: 'Dusk Foraging', binding: 'Hearth Binding' }[id] || id;
}

function r2(n) { return Math.round(n * 100) / 100; }

export function giveStarter(player) {
  player.equip.weapon = 'worn_dirk';
  player.equip.armor = 'cloth_wrap';
  invAdd(player.inv, 'hearthbread', 3);
  invAdd(player.inv, 'ember_coin', 80);
  invAdd(player.inv, 'tutor_note', 1);
  player.refresh();
  player.hp = player.maxHp;
}

export { serializeInv, EQUIP_SLOTS, ACTION };

File: py/player.py
"""Authoritative player: stats, pack, bank, equipment, skills."""
from __future__ import annotations

from constants import SPAWN, INV_SIZE, BANK_SIZE, EQUIP_SLOTS, uid, ACTION, xp_to_level, hue_from_name, now
from protocol import pack, S2C
from catalog import item, player_stats, skill_level
from systems.inventory import empty_inv, add_item as inv_add, remove_item as inv_remove, count_of, can_fit, serialize_inv
from netutil import send_ws


class Player:
    def __init__(self, ws, name):
        self.id = uid("p")
        self.ws = ws
        self.name = name
        self.x = SPAWN["x"]
        self.y = SPAWN["y"]
        self.dir = 0
        self.path = []
        self.hp = 40
        self.maxHp = 40
        self.inv = empty_inv(INV_SIZE)
        self.bank = empty_inv(BANK_SIZE)
        self.equip = {"weapon": None, "armor": None, "charm": None}
        self.skills = {"combat": {"xp": 0}, "foraging": {"xp": 0}, "binding": {"xp": 0}}
        self.action = ACTION["IDLE"]
        self.target = None
        self.tradeId = None
        self.tradeAsk = None
        self.tradeAskTo = None
        self.intent = None
        self.channel = None
        self.buff = None
        self.lastSwing = 0
        self.nextSwing = 0
        self.atk = 4
        self.def_ = 2
        self.joinedAt = now()
        self.lastChat = 0
        self.regenAcc = 0
        self.hue = hue_from_name(name)
        self.deadFor = 0
        self.talking = None
        self.talkingId = None

    def send(self, typ, payload=None):
        send_ws(self.ws, pack(typ, payload or {}))

    def add_item(self, iid, qty):
        return inv_add(self.inv, iid, qty)

    def remove_item(self, iid, qty):
        return inv_remove(self.inv, iid, qty)

    def count_item(self, iid):
        return count_of(self.inv, iid)

    def has_items(self, lst):
        for row in lst:
            if count_of(self.inv, row["id"]) < row["qty"]:
                return False
        return True

    def can_fit(self, iid, qty):
        return can_fit(self.inv, iid, qty)

    def bonus_atk(self):
        n = 0
        for slot in EQUIP_SLOTS:
            it = item(self.equip[slot]) if self.equip.get(slot) else None
            if it:
                n += it.get("atk") or 0
        return n

    def bonus_def(self):
        n = 0
        for slot in EQUIP_SLOTS:
            it = item(self.equip[slot]) if self.equip.get(slot) else None
            if it:
                n += it.get("def") or 0
        return n

    def bonus_hp(self):
        n = 0
        for slot in EQUIP_SLOTS:
            it = item(self.equip[slot]) if self.equip.get(slot) else None
            if it:
                n += it.get("hp") or 0
        return n

    def refresh(self):
        s = player_stats(self)
        ratio = self.hp / self.maxHp if self.maxHp > 0 else 1
        self.maxHp = s["maxHp"]
        self.atk = s["atk"]
        self.def_ = s["def"]
        self.hp = max(0, min(s["maxHp"], round(ratio * s["maxHp"])))

    def add_xp(self, skill_id, amount):
        if skill_id not in self.skills:
            self.skills[skill_id] = {"xp": 0}
        before = skill_level(self, skill_id)
        self.skills[skill_id]["xp"] += max(0, int(amount or 0))
        after = skill_level(self, skill_id)
        prog = xp_to_level(self.skills[skill_id]["xp"])
        self.send(S2C["XP"], {
            "skill": skill_id, "xp": self.skills[skill_id]["xp"], "amount": amount,
            "level": after, "into": prog["into"], "need": prog["need"],
        })
        if after > before:
            self.refresh()
            self.hp = min(self.maxHp, self.hp + 8)
            self.send(S2C["NOTIFY"], {"text": f"{label_skill(skill_id)} rises to {after}.", "kind": "xp"})
        return after

    def public_view(self):
        return {
            "id": self.id,
            "name": self.name,
            "x": r2(self.x),
            "y": r2(self.y),
            "hp": round(self.hp),
            "maxHp": self.maxHp,
            "action": self.action if isinstance(self.action, str) else (self.action or {}).get("type") if isinstance(self.action, dict) else "idle",
            "dir": r2(self.dir),
        }

    def private_view(self):
        pv = self.public_view()
        pv.update({
            "inv": serialize_inv(self.inv),
            "bank": serialize_inv(self.bank),
            "equip": {"weapon": self.equip.get("weapon"), "armor": self.equip.get("armor"), "charm": self.equip.get("charm")},
            "skills": {
                "combat": self.skills["combat"]["xp"],
                "foraging": self.skills["foraging"]["xp"],
                "binding": self.skills["binding"]["xp"],
            },
            "atk": self.atk,
            "def": self.def_,
            "target": self.target,
            "coins": self.count_item("ember_coin"),
            "channel": {"type": self.channel["type"], "until": self.channel["until"]} if self.channel else None,
        })
        return pv


def label_skill(sid):
    return {"combat": "Ashen Combat", "foraging": "Dusk Foraging", "binding": "Hearth Binding"}.get(sid, sid)


def r2(n):
    return round(n * 100) / 100


def give_starter(player):
    player.equip["weapon"] = "worn_dirk"
    player.equip["armor"] = "cloth_wrap"
    inv_add(player.inv, "hearthbread", 3)
    inv_add(player.inv, "ember_coin", 80)
    inv_add(player.inv, "tutor_note", 1)
    player.refresh()
    player.hp = player.maxHp
